app/controllers/candidate_interface/new_references/name_controller.rb

```diff
@@ -0,0 +1,55 @@
+module CandidateInterface
+  module NewReferences
+    class NameController < BaseController
+      before_action :verify_name_is_editable, only: %i[new create]
+      before_action :redirect_to_review_page_unless_reference_is_editable, :set_edit_backlink, only: %i[edit update]
+
+      def new
+        @reference_name_form = Reference::RefereeNameForm.build_from_reference(@reference)
+      end
+
+      def create
+        @reference_name_form = Reference::RefereeNameForm.new(referee_name_param)
+
+        if @reference_name_form.save(current_application, params[:referee_type], reference: @reference)
+          redirect_to candidate_interface_new_references_email_address_path(@reference&.id || current_application.application_references.last.id)
+        else
+          track_validation_error(@reference_name_form)
+          render :new
+        end
+      end
+
+      def edit
+        @reference_name_form = Reference::RefereeNameForm.build_from_reference(@reference)
+      end
+
+      def update
+        @reference_name_form = Reference::RefereeNameForm.new(referee_name_param)
+
+        if @reference_name_form.update(@reference)
```

Comment:
to reduce the logical complexity a bit, how about
```
        if @reference_name_form.update(@reference)
          redirect_to reference_name_form_return_path
        else
          track_validation_error(@reference_name_form)
          render :edit
        end
```
with a private method
```
      def reference_name_form_return_path
        return_to_path.presence || references_review_unsubmitted_path(@reference.id)
      end
```

Reply:
incidentally... [yes lawd](https://api.rubyonrails.org/classes/Object.html#method-i-presence)

Reply:
actually, `presence` changes the game. can get rid of the nested logic without needing a private method:

`redirect_to return_to_path.presence || references_review_unsubmitted_path(@reference.id)`